### ISSUE TITLE
BakeSuperMacaroon on startup if it does not exist

### ISSUE THREAD

**Issue by ZZiigguurraatt** [lightninglabs/lightning-terminal#910]
**Is your feature request related to a problem? Please describe.**
Would like an easier way to bake a super macaroon.

**Describe the solution you'd like**
Add a config option to litd that will allow a super macaroon to be baked automatically on startup if it does not already exist.

**Describe alternatives you've considered**
Use the gRPC or CLI (https://lightning.engineering/api-docs/api/lit/proxy/bake-super-macaroon/) to do it manually after first creating the node. This is annoying because the user needs to coordinate a one time task to determine whether the macaroon exists or not and deal with yet another macaroon to control litd to then create the super macaroon.

**Additional context**
When using taproot assets, you often need to use both lnd and tapd RPC's and having to manage multiple connections to the same litd instance with multiple macaroons is a nuisance.

When creating docker containers for litd, normally the one time tasks of doing container setup are done in the Dockerfile, but litd does not normally start and create the node data until after the docker container is started and appropriate volumes are mounted. With this paradigm in mind, all other macaroons are automatically created on startup if they do not exist, and the super macaroon is an exception that I don't think we should have to deal with.

**Comment by ZZiigguurraatt** [comment]
Related: I noticed that `litcli ln` now seems to bake it's own super macaroon in some way. Is this right? Does it create a new super macaroon every time the CLI is used? With `litcli ln` calls I seem to be able to use them without passing the tapd macaroon. I also saw some error in the log file if I pass the wrong macaroon, it says it can't bake a super macaroon. However, I don't see where it is writing a super macaroon to disk anywhere, so maybe it just discards it at the end of the call. Any idea what `litcli ln` is doing?

**Comment by ellemouton** [comment]
cc @guggero  for the question regarding what `litcli ln` does with respect to super macaroon baking

**Comment by guggero** [comment]
Yes, we use [`connectSuperMacClient`](https://github.com/lightninglabs/lightning-terminal/blob/023c91f908cdcd6d8412b2483a7536f9b2dbfde8/cmd/litcli/main.go#L302) to create a super macaroon on the fly just for the connection to `tapd`. We can do that if we have access to the `lit.macaroon` which has that capability.
Though with some of the latest changes, you should also be able to create a super macaroon using the `admin.macaroon` from `lnd`.

**Comment by Cyberguru1** [comment]
After reviewing how the proxy service  `BakeSuperMacaroon` works, this is a proposal to implement the auto-baking of the super macaroon on startup:

####  Configuration Flags
Three configuration options will be added:
* `--bake-super-macaroon` (boolean): Enables the startup baking logic (defaults to `false`).
* `--super-macaroon-read-only` (boolean): Restricts the baked macaroon to read-only permissions (defaults to `false` for read-write permissions).
* `--super-macaroon-path` (string, optional): Specifies the output path. Defaults to `<lit-dir>/<network>/super.macaroon`.


####  Startup Baking Logic and Integration Tests
* **Auto-bake Logic**: During initialization, if `--bake-super-macaroon` is specified, the server will check if a macaroon already exists at the configured path. If it does not, a new super macaroon will be baked. The permissions will be filtered based on the `--super-macaroon-read-only` toggle (using the permissions manager's `ActivePermissions` API), and the resulting macaroon will be written to disk.
* **Integration Tests**: An integration test will be added to test the options.


Let me know what you think or if you have any suggestions on it. Thanks.

cc @ViktorT-11

**Comment by ViktorT-11** [comment]
Thanks for looking into this @Cyberguru1!

I'm adding some suggestions for your comments below:
> --bake-super-macaroon (boolean): Enables the startup baking logic (defaults to false).
--super-macaroon-read-only (boolean): Restricts the baked macaroon to read-only permissions (defaults to false for read-write permissions).

I think instead of using 2 booleans for this, you could add just add the `bake-super-macaroon` option and make that `string`, and make that `choice` config option. You can then add `none` as the default choice, & `readonly` or `read-write` as the other choices. If you have a better suggestion for the name any of the choices, feel free to change this as you'd prefer. 

  > Startup Baking Logic and Integration Tests
Auto-bake Logic: During initialization, if --bake-super-macaroon is specified, the server will check if a macaroon already exists at the configured path. If it does not, a new super macaroon will be baked. The permissions will be filtered based on the --super-macaroon-read-only toggle (using the permissions manager's ActivePermissions API), and the resulting macaroon will be written to disk.
Integration Tests: An integration test will be added to test the options.

So one important caveat for how `litcli bakesupermacaroon` works today is that if for example `loop` is disabled by setting the `loop-mode` config flag to `disable`, the generated macaroon from `litcli bakesupermacaroon` will not contain any `loop` permissions. We'd want the auto generated `supermacaroon` from this new functionality to do the same thing.

Since the actual permissions the `litcli bakesupermacaroon` command will generate differs over time, we'd want to keep the macaroon at the specified path up to date. Therefore, in addition to your idea of "if a macaroon already exists at the configured path", you'd also want to ensure that if a macaroon does already exist at that path, it should have also contain the same permissions which a macaroon from `litcli bakesupermacaroon` command would include if executed during that specific startup of litd. If it doesn't the macaroon at the specified path should be overwritten with a new macaroon that contains the expected permissions.

For the normal `lit.macaroon`, this is already handled through the `lndclient.MacaroonService` which contains the logic to check the permissions of the `lit.macaroon` and overwrite the macaroon if the permissions differ. Those permissions are set here for the normal macaroon: https://github.com/lightninglabs/lightning-terminal/blob/3e5f36c0b86277c8034266fb091a53f7bd34a058/terminal.go#L1029 
Now since the `bakesupermacaroon` command instead makes an RPC call to `lnrpc.Lightning/BakeMacaroon`, you'd need to either make sure that functionality of `lndclient.MacaroonService` is exported and can be re-used here (i.e. open a PR for that in the `lndclient` repo if it is not exported today), or replicate that logic for `litd`. I'll let you choose which option you think is most suitable.

**Comment by ZZiigguurraatt** [comment]
> Since the actual permissions the `litcli bakesupermacaroon` command will generate differs over time, we'd want to keep the macaroon at the specified path up to date.

You mean when with a new release version?

**Comment by ZZiigguurraatt** [comment]
> I think instead of using 2 booleans for this, you could add just add the `bake-super-macaroon` option and make that `string`, and make that `choice` config option. You can then add `none` as the default choice, & `readonly` or `read-write` as the other choices. If you have a better suggestion for the name any of the choices, feel free to change this as you'd prefer.

I agree.

**Comment by ZZiigguurraatt** [comment]
> So one important caveat for how `litcli bakesupermacaroon` works today is that if for example `loop` is disabled by setting the `loop-mode` config flag to `disable`, the generated macaroon from `litcli bakesupermacaroon` will not contain any `loop` permissions. We'd want the auto generated `supermacaroon` from this new functionality to do the same thing.

Probably if we have loop enabled and then later disable it we should make a new super macaroon be baked without loop permissions that overwrites the old one that had loop permissions?

**Comment by ViktorT-11** [comment]
> You mean when with a new release version?

Yes, that can be one cause which changes the permissions, as new releases can include additional functionality which requires new permissions.

Alternatively, like you also mentioned, permissions for the `litcli bakesupermacaroon` changes over time when the config changes to `enable` & `disable` different sub-servers in `litd` (i.e. `loop`, `pool`, `tapd` etc).

> Probably if we have loop enabled and then later disable it we should make a new super macaroon be baked without loop permissions that overwrites the old one that had loop permissions?

Yes, that I agree that this should be the intended result.

**Comment by Cyberguru1** [comment]
Sorry for the late response.
@ViktorT-11 Thank you for looking into the proposal and the suggestions 🙏 


> I think instead of using 2 booleans for this, you could add just add the bake-super-macaroon option and make that string, and make that choice config option. You can then add none as the default choice, & readonly or read-write as the other choices. If you have a better suggestion for the name any of the choices, feel free to change this as you'd prefer.

Yes, makes sense. Will use the choice config option approach.

> Therefore, in addition to your idea of "if a macaroon already exists at the configured path", you'd also want to ensure that if a macaroon does already exist at that path, it should have also contain the same permissions which a macaroon from litcli bakesupermacaroon command would include if executed during that specific startup of litd. If it doesn't the macaroon at the specified path should be overwritten with a new macaroon that contains the expected permissions.

Agreed. Will update to ensure the macaroon stays in sync with the expected permissions on each startup.


> Now since the bakesupermacaroon command instead makes an RPC call to lnrpc.Lightning/BakeMacaroon, you'd need to either make sure that functionality of lndclient.MacaroonService is exported and can be re-used here (i.e. open a PR for that in the lndclient repo if it is not exported today), or replicate that logic for litd. I'll let you choose which option you think is most suitable.

I looked into `lndclient.MacaroonService` and found two relevant pieces. The exported [`lightningClient.CheckMacaroonPermissions`](https://github.com/lightninglabs/lndclient/blob/f3260ef33ca86362aa12b312f4bf9583c2037954/lightning_client.go#L4193) makes an RPC call to `lnrpc.Lightning/CheckMacaroonPermissions`, but as noted in the codebase this only performs a subset check on permissions, so a macaroon with stale extra permissions would still pass as valid. The unexported [`macaroonMatchesPermissions`](https://github.com/lightninglabs/lndclient/blob/f3260ef33ca86362aa12b312f4bf9583c2037954/macaroon_service.go#L411) does perform an exact match, which is what we need here.

I'll go with replicating the exact-match permission check logic inside `litd`.
